File: multiCamControllerPython/src/services/device_discovery.py
"""Device discovery via mDNS"""

from PySide6.QtCore import QObject, Signal as pyqtSignal, QTimer
from zeroconf import ServiceBrowser, ServiceListener, Zeroconf
from typing import Optional
import socket
import logging

from models.device import MultiCamDevice
from models.app_state import AppState
from utils.constants import SERVICE_TYPE, DISCOVERY_TIMEOUT

logger = logging.getLogger(__name__)


class MultiCamServiceListener(ServiceListener):
    """Listener for mDNS service events"""

    def __init__(self, discovery_service: "DeviceDiscovery"):
        self.discovery = discovery_service

    def add_service(self, zc: Zeroconf, type_: str, name: str):
        """Called when a service is discovered"""
        logger.info(f"Service added: {name}")
        info = zc.get_service_info(type_, name)
        if info:
            self.discovery.service_resolved(name, info)

    def remove_service(self, zc: Zeroconf, type_: str, name: str):
        """Called when a service is removed"""
        logger.info(f"Service removed: {name}")
        self.discovery.service_removed(name)

    def update_service(self, zc: Zeroconf, type_: str, name: str):
        """Called when a service is updated"""
        logger.info(f"Service updated: {name}")


class DeviceDiscovery(QObject):
    """Discovers devices on the local network via mDNS"""

    device_discovered = pyqtSignal(MultiCamDevice)
    device_removed = pyqtSignal(str)
    discovery_error = pyqtSignal(str)

    def __init__(self, app_state: AppState, parent=None):
        super().__init__(parent)
        self.app_state = app_state
        self.zeroconf: Optional[Zeroconf] = None
        self.browser: Optional[ServiceBrowser] = None
        self.timeout_timer = QTimer(self)
        self.timeout_timer.timeout.connect(self.on_discovery_timeout)

    def start_discovery(self):
        """Start discovering devices"""
        try:
            self.app_state.is_discovering = True
            self.app_state.update_status("Discovering devices...")
            self.app_state.clear_devices()

            logger.info(f"Starting mDNS discovery for {SERVICE_TYPE}")

            self.zeroconf = Zeroconf()
            listener = MultiCamServiceListener(self)
            self.browser = ServiceBrowser(self.zeroconf, SERVICE_TYPE, listener)

            # Start timeout timer
            self.timeout_timer.start(DISCOVERY_TIMEOUT * 1000)

        except Exception as e:
            logger.error(f"Discovery failed: {e}")
            self.app_state.update_status(f"Discovery failed: {str(e)}")
            self.app_state.is_discovering = False
            self.discovery_error.emit(str(e))

    def stop_discovery(self):
        """Stop discovering devices"""
        logger.info("Stopping mDNS discovery")
        self.timeout_timer.stop()

        if self.browser:
            self.browser.cancel()
            self.browser = None

        if self.zeroconf:
            self.zeroconf.close()
            self.zeroconf = None

        self.app_state.is_discovering = False

    def service_resolved(self, name: str, info):
        """Called when a service is resolved"""
        try:
            # Extract IP address
            if info.addresses:
                ip = socket.inet_ntoa(info.addresses[0])
            else:
                logger.warning(f"No address for service {name}")
                return

            port = info.port

            device = MultiCamDevice(name=name, ip=ip, port=port, service_type=SERVICE_TYPE)

            logger.info(f"Found device: {device.display_name} at {ip}:{port}")
            self.app_state.add_device(device)
            self.device_discovered.emit(device)

        except Exception as e:
            logger.error(f"Failed to process service {name}: {e}")

    def service_removed(self, name: str):
        """Called when a service is removed"""
        logger.info(f"Device removed: {name}")
        self.app_state.remove_device(name)
        self.device_removed.emit(name)

    def on_discovery_timeout(self):
        """Handle discovery timeout"""
        self.timeout_timer.stop()

        device_count = len(self.app_state.discovered_devices)
        if device_count > 0:
            self.app_state.update_status(f"Found {device_count} device(s)")
        else:
            logger.warning("Discovery timeout - no devices found")
            self.app_state.update_status(
                "Auto-discovery timed out. Try manual device connection."
            )

        self.stop_discovery()

    def add_manual_device(self, ip: str, port: int, name: Optional[str] = None):
        """Manually add a device by IP and port"""
        if name is None:
            name = f"manual-{ip}"

        device = MultiCamDevice(name=name, ip=ip, port=port, service_type="_manual._tcp.")

        logger.info(f"Manually added device: {device.display_name} at {ip}:{port}")
        self.app_state.add_device(device)
        self.device_discovered.emit(device)

File: multiCamControllerMacos/multiCamControllerMacos/Services/DeviceDiscovery.swift
//
//  DeviceDiscovery.swift
//  multiCamControllerMacos
//
//  Created by Claude Code on 9/25/25.
//

import Foundation
import Network

@MainActor
class DeviceDiscovery: NSObject, ObservableObject {
    private let serviceBrowser = NetServiceBrowser()
    private let serviceType = "_multicam._tcp."
    private let serviceDomain = ""

    @Published var discoveredServices: [NetService] = []

    weak var appState: AppState?

    override init() {
        super.init()
        serviceBrowser.delegate = self
    }

    func startDiscovery() {
        appState?.isDiscovering = true
        appState?.updateStatus("🔍 Discovering devices...")
        appState?.clearDevices()

        print("🔍 Starting mDNS discovery for \(serviceType)")
        serviceBrowser.searchForServices(ofType: serviceType, inDomain: serviceDomain)

        // Set a timeout to automatically retry or show alternative options
        DispatchQueue.main.asyncAfter(deadline: .now() + 10.0) {
            if let appState = self.appState, appState.isDiscovering && appState.discoveredDevices.isEmpty {
                print("⚠️ mDNS discovery timeout - suggesting manual device entry")
                appState.updateStatus("⚠️ Auto-discovery timed out. Try manual device connection.")
                appState.isDiscovering = false
                self.stopDiscovery()
            }
        }
    }

    func stopDiscovery() {
        appState?.isDiscovering = false
        serviceBrowser.stop()
        print("🛑 Stopped mDNS discovery")
    }

    private func createDevice(from service: NetService) -> MultiCamDevice? {
        // For immediate device creation, we can use the hostname if available
        // The service might not be fully resolved yet, but we can still create the device

        var deviceIP = service.hostName ?? "unknown"

        // Try to extract IP from addresses if available
        if let addresses = service.addresses, !addresses.isEmpty {
            let address = addresses[0]
            var hostname = [CChar](repeating: 0, count: Int(NI_MAXHOST))

            let result = getnameinfo(
                address.withUnsafeBytes { $0.bindMemory(to: sockaddr.self).baseAddress! },
                socklen_t(address.count),
                &hostname, socklen_t(hostname.count),
                nil, 0,
                NI_NUMERICHOST
            )

            if result == 0, let ip = String(cString: hostname, encoding: .utf8) {
                deviceIP = ip
            }
        } else if let hostName = service.hostName {
            // Use hostname if addresses not available yet
            deviceIP = hostName.replacingOccurrences(of: ".local.", with: "")
        }

        // Validate service port (NetService.port can return -1 if unresolved)
        let validPort = service.port > 0 ? service.port : 8080

        let device = MultiCamDevice(
            name: service.name,
            ip: deviceIP,
            port: validPort,
            serviceType: service.type
        )

        print("📱 Found device: \(device.displayName) at \(deviceIP):\(validPort)")
        return device
    }
}

// MARK: - NetServiceBrowserDelegate
extension DeviceDiscovery: NetServiceBrowserDelegate {
    nonisolated func netServiceBrowser(_ browser: NetServiceBrowser, didFind service: NetService, moreComing: Bool) {
        Task { @MainActor in
            print("🔍 Found service: \(service.name)")
            discoveredServices.append(service)

            // Start resolving the service to get proper IP address
            service.delegate = self
            service.resolve(withTimeout: 5.0)

            if !moreComing {
                let deviceCount = appState?.discoveredDevices.count ?? 0
                if deviceCount > 0 {
                    appState?.updateStatus("✅ Found \(deviceCount) device(s)")
                } else {
                    appState?.updateStatus("Waiting")
                }
                appState?.isDiscovering = false
                stopDiscovery()
            }
        }
    }

    nonisolated func netServiceBrowser(_ browser: NetServiceBrowser, didRemove service: NetService, moreComing: Bool) {
        Task { @MainActor in
            print("❌ Service removed: \(service.name)")
            discoveredServices.removeAll { $0.name == service.name }
            appState?.removeDevice(named: service.name)
        }
    }

    nonisolated func netServiceBrowser(_ browser: NetServiceBrowser, didNotSearch errorDict: [String: NSNumber]) {
        Task { @MainActor in
            let errorCode = errorDict["NSNetServicesErrorCode"]?.intValue ?? 0
            let errorDomain = errorDict["NSNetServicesErrorDomain"]?.description ?? "Unknown"

            print("❌ Service browser did not search - Domain: \(errorDomain), Code: \(errorCode)")
            print("❌ Full error dict: \(errorDict)")

            var errorMessage = "Discovery failed"
            if errorCode == -72000 {
                errorMessage = "Network permission denied. Please check app entitlements and network settings."
            } else if errorCode == -72004 {
                errorMessage = "Network service not found or unavailable."
            }

            appState?.updateStatus("❌ \(errorMessage)")
            appState?.isDiscovering = false
        }
    }

    nonisolated func netServiceBrowserWillSearch(_ browser: NetServiceBrowser) {
        print("🔍 Service browser will search")
    }

    nonisolated func netServiceBrowserDidStopSearch(_ browser: NetServiceBrowser) {
        Task { @MainActor in
            print("🛑 Service browser stopped search")
            appState?.isDiscovering = false
        }
    }
}

// MARK: - NetServiceDelegate
extension DeviceDiscovery: NetServiceDelegate {
    nonisolated func netServiceDidResolveAddress(_ sender: NetService) {
        Task { @MainActor in
            print("✅ Service resolved: \(sender.name)")
            if let device = createDevice(from: sender) {
                appState?.addDevice(device)
                print("📱 Added resolved device: \(device.displayName) at \(device.ip):\(device.port)")
            }
        }
    }

    nonisolated func netService(_ sender: NetService, didNotResolve errorDict: [String: NSNumber]) {
        Task { @MainActor in
            print("❌ Service resolution failed for \(sender.name): \(errorDict)")
            // Still try to create device with default port as fallback
            if let device = createDevice(from: sender) {
                appState?.addDevice(device)
                print("📱 Added unresolved device: \(device.displayName) at \(device.ip):\(device.port)")
            }
        }
    }
}